File: chronoid-ts/src/AsyncSnowflakeGenerator.ts
import { SnowflakeId } from './SnowflakeId.js';
import { AsyncExhaustionStrategy } from './types.js';
import { SequenceExhausted } from './errors.js';

/**
 * Stateful generator for asynchronous contexts.
 */
export class AsyncSnowflakeGenerator {
  public last_timestamp: number = -1;
  public sequence: number = 0;

  // Time Cache for massive performance optimization natively avoiding Date objects
  private cachedPrefix: bigint = 0n;
  private cachedNextYearMs: number = 0;
  private cachedStartOfYearMs: number = 0;
  private cachedYearOffset: bigint = 0n;

  /**
   * Internal constructor. Use SnowflakeGenerator.create() instead.
   */
  constructor(
    protected readonly base_year: number,
    protected readonly node_id: number,
    protected readonly worker_id: number,
    protected readonly strategy: AsyncExhaustionStrategy
  ) {}

  /**
   * Core synchronous generation engine shared natively between Async and Sync generators.
   * Pulls the absolute clock, ensures synchronization bounds, packs, and shifts the result.
   */
  public tryGenerate(nowMs: number = Date.now()): SnowflakeId | 'EXHAUSTED' {
    if (nowMs < this.last_timestamp) {
      throw new Error(`Clock moved backwards. Refusing to generate id for ${this.last_timestamp - nowMs} milliseconds`);
    }

    if (nowMs === this.last_timestamp) {
      if (this.sequence >= 1023) {
        return 'EXHAUSTED';
      }
      this.sequence++;
      // Zero allocations! Instantly bitwise append sequence to heavily cached prefix limits
      return SnowflakeId.from(BigInt(this.sequence) | this.cachedPrefix);
    }

    this.sequence = 0;
    this.last_timestamp = nowMs;

    // Update Year boundaries only when physically crossing the boundary
    if (nowMs >= this.cachedNextYearMs) {
      const d = new Date(nowMs);
      const yr = d.getUTCFullYear();
      this.cachedYearOffset = BigInt(yr - this.base_year) & 0xFFn;
      this.cachedStartOfYearMs = Date.UTC(yr, 0, 1);
      this.cachedNextYearMs = Date.UTC(yr + 1, 0, 1);
    }

    // Natively extract boundaries using pure modulo math without any object bindings spanning logic
    const day = BigInt(Math.floor((nowMs - this.cachedStartOfYearMs) / 86400000));
    const minute = BigInt(Math.floor(nowMs / 60000) % 1440);
    const millisecond = BigInt(nowMs % 60000);
    const node = BigInt(this.node_id);
    const worker = BigInt(this.worker_id);

    // Cache the completely shifted timestamp block mapping exclusively over sequence limit mappings natively.
    this.cachedPrefix = (worker << 10n) 
                      | (node << 14n) 
                      | (millisecond << 19n) 
                      | (minute << 35n) 
                      | (day << 46n) 
                      | (this.cachedYearOffset << 55n);

    // Sequence is 0 right now, returning inherently zero logic checks mapped naturally over the footprint.
    return SnowflakeId.from(this.cachedPrefix);
  }

  public async generate(): Promise<SnowflakeId> {
    while (true) {
      let nowMs = Date.now();
      const result = this.tryGenerate(nowMs);
      
      if (result !== 'EXHAUSTED') {
        return result;
      }

      // Exhaustion fallback strategies natively scoped for async implementation
      if (this.strategy === AsyncExhaustionStrategy.Throw) {
        throw new SequenceExhausted("Sequence exhausted for current millisecond.");
      } else {
        // WaitAsync: yield the event loop dynamically using setImmediate
        // This spins drastically faster than setTimeout (which is natively bound to 1-15ms OS ticks), 
        // allowing it to precisely capture the exact microsecond the clock rolls over!
        while (nowMs <= this.last_timestamp) {
          await new Promise(resolve => setImmediate(resolve));
          nowMs = Date.now();
        }
      }
    }
  }
}

File: chronoid-java/src/main/java/io/github/datakore/chronoid/AsyncSnowflakeGenerator.java
package io.github.datakore.chronoid;

import java.time.LocalDateTime;
import java.time.ZoneOffset;
import java.time.format.DateTimeFormatter;
import java.util.concurrent.CompletableFuture;

public class AsyncSnowflakeGenerator {
    private final int baseYear;
    private final int nodeId;
    private final int workerId;
    private final AsyncExhaustionStrategy strategy;

    long lastTimestamp = -1;
    int sequence = 0;

    // Caching
    private long cachedPrefix = 0;
    private long cachedNextYearMs = 0;
    private long cachedStartOfYearMs = 0;
    private long cachedYearOffset = 0;

    public AsyncSnowflakeGenerator(int baseYear, int nodeId, int workerId, AsyncExhaustionStrategy strategy) {
        this.baseYear = baseYear;
        this.nodeId = nodeId;
        this.workerId = workerId;
        this.strategy = strategy;
    }

    public synchronized SnowflakeId tryGenerate(long nowMs) {
        if (nowMs < lastTimestamp) {
            throw new ChronoidException.InvalidId("Clock moved backwards. Refusing to generate for " + (lastTimestamp - nowMs) + "ms");
        }

        if (nowMs == lastTimestamp && lastTimestamp != -1) {
            if (sequence >= 1023) {
                return null; // EXHAUSTED
            }
            sequence++;
            return new SnowflakeId(((long) sequence) | cachedPrefix);
        }

        // New millisecond
        sequence = 0;
        lastTimestamp = nowMs;

        if (nowMs >= cachedNextYearMs) {
            LocalDateTime dt = LocalDateTime.ofEpochSecond(nowMs / 1000, (int) ((nowMs % 1000) * 1_000_000), ZoneOffset.UTC);
            int yr = dt.getYear();
            cachedYearOffset = ((long) (yr - baseYear) & 0xFFL) << 55;

            LocalDateTime start = LocalDateTime.of(yr, 1, 1, 0, 0, 0);
            LocalDateTime end = LocalDateTime.of(yr + 1, 1, 1, 0, 0, 0);
            cachedStartOfYearMs = start.toInstant(ZoneOffset.UTC).toEpochMilli();
            cachedNextYearMs = end.toInstant(ZoneOffset.UTC).toEpochMilli();
        }

        long day = (nowMs - cachedStartOfYearMs) / 86400000L;
        long minute = (nowMs / 60000L) % 1440L;
        long millisecond = nowMs % 60000L;
        long node = (long) nodeId << 14;
        long worker = (long) workerId << 10;

        cachedPrefix = worker 
                     | node 
                     | (millisecond << 19) 
                     | (minute << 35) 
                     | (day << 46) 
                     | cachedYearOffset;

        return new SnowflakeId(cachedPrefix);
    }

    /**
     * Primary generation engine with high-precision yielding.
     * This method is the unified polling loop for both Async and Sync paths.
     */
    private SnowflakeId internalGenerate() {
        while (true) {
            long nowMs = System.currentTimeMillis();
            SnowflakeId id = tryGenerate(nowMs);
            if (id != null) return id;

            if (strategy == AsyncExhaustionStrategy.THROW) {
                throw new ChronoidException.SequenceExhausted();
            }

            // High-precision non-blocking wait until the next millisecond boundary
            long targetMs = lastTimestamp + 1;
            while (System.currentTimeMillis() < targetMs) {
                // Hint to the CPU that we are in a tight spin loop for micro-latency
                Thread.onSpinWait();
            }
        }
    }

    public CompletableFuture<SnowflakeId> generate() {
        return CompletableFuture.supplyAsync(this::internalGenerate);
    }

    /**
     * Optimized blocking generator for synchronous performance.
     */
    public SnowflakeId generateSync() {
        return internalGenerate();
    }
}
